site/app/controllers/admin/UsersController.php

```diff
@@ -928,50 +928,109 @@ public function uploadUserList($list_type = "classlist") {
-                        $bad_rows[] = ($row_num + 1);
-                        break;
+            // Bounds check to ensure minimum required number of rows is present.
+            if (!count($vals) >= 5) {
+                $bad_row_details[$row_num + 1][] = 'column Count';
+                if (!in_array('column_count', $bad_columns)) {
+                    $bad_columns[] = 'column_count';
+                }
+            }
+            // Username must contain only lowercase alpha, numbers, underscores, hyphens
+            if (!User::validateUserData('user_id', $vals[0])) {
+                $bad_row_details[$row_num + 1][] = 'user_id';
+                if (!in_array('user_id', $bad_columns)) {
+                    $bad_columns[] = 'user_id';
+                }
+            }
+            // First Name must be alpha characters, white-space, or certain punctuation.
+            if (!User::validateUserData('user_legal_firstname', $vals[1])) {
+                $bad_row_details[$row_num + 1][] = 'first name';
+                if (!in_array('user_legal_firstname', $bad_columns)) {
+                    $bad_columns[] = 'user_legal_firstname';
+                }
+            }
+            // Last Name must be alpha characters, white-space, or certain punctuation.
+            if (!User::validateUserData('user_legal_lastname', $vals[2])) {
+                $bad_row_details[$row_num + 1][] = 'last Name';
+                if (!in_array('user_legal_lastname', $bad_columns)) {
+                    $bad_columns[] = 'user_legal_lastname';
+                }
+            }
+            // Check email address for appropriate format. e.g. "student@university.edu", "student@cs.university.edu", etc.
+            if (!User::validateUserData('user_email', $vals[3])) {
+                $bad_row_details[$row_num + 1][] = 'user email';
+                if (!in_array('user_email', $bad_columns)) {
+                    $bad_columns[] = 'user_email';
+                }
+            }
+            /* $row[4] validation varies by $list_type
+               "classlist" validates registration_section, and "graderlist" validates user_group */
+            if (!$row4_validation_function()) {
+                $bad_row_details[$row_num + 1][] = $list_type === 'classlist'
+                    ? 'Registration section'
+                    : 'Grader-group';
+                if (!in_array('row4_validation', $bad_columns)) {
+                    $bad_columns[] = 'row4_validation';
+                }
+            }
+            /* Database password cannot be blank, no check on format.
+               Automatically validate if NOT using database authentication (e.g. using PAM authentication) */
+            if (!(!$use_database || User::validateUserData('user_password', $vals[5]))) {
+                $bad_row_details[$row_num + 1][] = 'User password';
+                if (!in_array('user_password', $bad_columns)) {
+                    $bad_columns[] = 'user_password';
```

Comment:
I was able to recreate the errors. They look great and are a huge improvement from before. I just didn't know to recreate this since the upload class list doesn't call for a password.